functional_tests/README.md

```diff
@@ -0,0 +1,303 @@
+> @step('a thing happens')
+> def step_to_do_a_thing(context: Context):
+>     context.a_new_context_var = helper_function(context.my_scenario_data)
+>     ...
+>
+> def helper_function(my_scenario_data):
+>     ...
+>     return new_data
+> ```
+
+> [!CAUTION]
+>
+> **Not this:**
+>
+> ```python
+> @step('a thing happens')
+> def step_to_do_a_thing(context: Context):
+>     helper_function_which_overwrites_context(context)
+>     ...
+>
+> def helper_function_which_overwrites_context(context: Context) -> None:
+>     ...
+>     context.a_new_context_var = new_data
+> ```
+
+### Sharing Code Between Steps
+
+Step files should not import code from other step files, where code can be shared between steps they should either be in
+the same file, or the shared code should be factored out into the [step_helpers](step_helpers) module.
+
+This is to avoid potential circular imports and make it clear which code is specific to certain steps, and which is
+reusable across any steps.
+
+### Step wording
+
+Steps should be written in full and concise sentences, avoiding unnecessary abbreviations and shorthand. They should be
+as understandable and as non-technical as possible.
+
+### Assertions
+
+Assertions should use the [Playwright assertions](https://playwright.dev/docs/test-assertions) wherever possible. This
+has built in retry and timeout logic where applicable, to allow grace periods for content to load on the page, so will
+be more robust than attempting to retrieve data from the page and make plain python assertions.
+
+### Step parameter types
+
+Where we need [step parameters](https://behave.readthedocs.io/en/stable/tutorial.html#step-parameters) to include more
+complex data than single strings or the other basic types supported by the default parser, we
+use [custom registered types](https://behave.readthedocs.io/en/stable/api.html#behave.register_type). These are
+registered in the [environment.py](environment.py) so they are available to all steps.
+
+## How the Tests Work
+
+### Django Test Runner and Test Case
+
+Due to issues with the Django `TransactionTestCase` which prevent us using the built-in database teardown/setup
+in between scenarios, we have implemented our own database snapshot and restore pattern between tests. We still make use
+of the Django test case, specifically the
+[`LiveServerTestCase`](https://docs.djangoproject.com/en/stable/topics/testing/tools/#django.test.LiveServerTestCase)
+to perform the initial database setup and run a live server on a random port for the tests.
+
+### Database Snapshot and Restore
+
+We are using [DSLR](https://pypi.org/project/dslr/) for fast database snapshots and restores.
+
+After we have used the Django test runner to set up the test database and initialise it by running migrations, we take a
+DSLR snapshot of this clean, initial state. In the test case fixture, post test, we then restore the clean snapshot,
+ensuring each test gets a clean, migrated database, isolated from other tests.
```

Comment:
Question: Django should be handling this itself with its test classes, using a test connection? Is that not working?

Reply:
There is more detail about this is in the [why-arent-we-using-existing-django-testing-modules section](https://github.com/ONSdigital/dis-wagtail/blob/308d61bb4c32593d2fa446f2cd633f15520c97de/functional_tests/README.md#why-arent-we-using-existing-django-testing-modules). TL;DR, we're still using the test runner to set up the test database initially, but the between test teardown was breaking things, removing the required data which is seeded in migrations.

Reply:
I suspect the solution is https://github.com/ONSdigital/dis-wagtail/pull/31#discussion_r1875705404, but as Dan mentions it's probably one to revisit.

Reply:
Yes, I think a combination of `serialised_rollback` and the fix in https://github.com/wagtail/wagtail/pull/12649 would resolve it. I think the core issue was the combination of data flushing and the Wagtail not being able to cope with restoring from scratch with fixtures. However with a quick play copying the solution in our `TransactionTestCase` I still wasn't able to get it passing locally.

I would agree with Dan that we should continue with the current solution and revisit.